text/0000-default-field-values.md

```diff
@@ -0,0 +1,1658 @@
+`PhantomData` carries no useful information since it is a ZST.
+Therefore, the result of having to add a field of type `PhantomData`
+is a regression to ergonomics throughout. With default values you can
+improve the situation slightly and provide the value in a single place:
+
+```rust
+use std::marker::PhantomData;
+
+struct Vec<T> {
+    data: *const T, // *const for variance!
+    len: usize,
+    cap: usize,
+    _marker: PhantomData<T> = PhantomData,
+}
+```
+
+You then go on to define some operations for `Vec<T>`:
+
+```rust
+impl<T> Vec<T> {
+    pub fn new() -> Self {
+        Self {
+            // other fields...
+            ..
+        }
+    }
+
+    pub fn with_capacity(capacity: usize) -> Self {
+        Self {
+            // other fields...
+            ..
+        }
+    }
+}
+```
+
+### Defaults for `enum`s
+
+So far, you've have seen how ergonomics were improved for `struct`s.
+For `enum`s, FRU cannot be used as for structs and neither can
+`#[derive(Default)]` be used for each variant or the entire enum.
+Instead, at best, a constructor function can be offered for each variant.
+For a large enum which for example encodes an expression language such
+this could become quite a lot of variants.
+
+By allowing each individual variant of an `enum` to have default values,
+a more profound can achieve a more profound improvement to ergonomics
+than for `struct`s. For example, you may write:
+
+```rust
+enum Foo<'a> {
+    Bar {
+        alpha: u8 = 42,
+        beta: &'a str = "beta's default value",
+    },
+    Baz {
+        gamma: Vec<u8> = Vec::new(),
+        delta: f32
+    }
+}
+```
+
+and then these defaults may be used like so:
+
+```rust
+let a = Foo::Bar { .. };
+let b = Foo::Bar { alpha: 1, .. };
+let c = Foo::Bar { beta: "another beta", .. };
+
+let d = Foo::Baz { delta: 1.0, .. };
+```
+
+## `#[derive(Default)]` in more cases
+
+The `#[derive(..)]` ("custom derive") mechanism works by defining procedural
+*macros*. Because they are macros, these operate on abstract *syntax* and
+don't have more information available. Therefore, when you `#[derive(Default)]`
+on a data type definition as with:
+
+```rust
+#[derive(Default)]
+struct Foo {
+    bar: u8,
+    baz: String,
+}
+```
+
+it only has the immediate "textual" definition available to it.
+
+Because Rust currently does not have an in-language way to define default values,
+you cannot `#[derive(Default)]` in the cases where you are not happy with the
+natural default values that each field's type provides. By extending the syntax
+of Rust such that default values can be provided, `#[derive(Default)]` can be
+used in many more circumstances and thus boilerplate is further reduced.
+
+Currently, `#[derive(Default)]` is not usable for `enum`s. To rectify this
+situation and to allow enum variants to take advantage of their newfound
+ability to have default field values, a `#[default]` attribute is introduce
```

Comment:
This feels like a good thing to have, but it's probably worth mentioning that it's orthogonal to the rest of the RFC.